modules/contact/src/actions/ContactAction.C

```diff
@@ -1028,6 +1060,120 @@ ContactAction::addNodeFaceContact()
+            "1) Those sidesets wrap around the surfaces of the bodies, 2) Axisymmetry of the "
+            "geometry and the mesh exists. Deviations from axisymmetry may give rise to wrong "
+            "contact pairing. The user is encouraged to check the screen output to confirm pairs "
+            "were assigned correctly.");
+
+  _mesh = &_problem->mesh();
+
+  if (!_mesh)
+    mooseError("Internal error in contact action.");
+
+  if (!_mesh->getMesh().is_serial())
+    paramError(
+        "nodeset_list",
+        "The generation of automatic contact pairs in the contact action requires a serial mesh.");
+
+  // Compute centers of gravity for each nodeset
+  std::vector<std::pair<BoundaryName, Point>> nodeset_list_cog;
+
+  const auto & nodeset_ids = _mesh->meshNodesetIds();
+
+  // Loop over nodesets
+  for (const auto & nodeset_id : _nodeset_list)
+  {
+    // If the nodeset provided in the input file isn't in the mesh, error out.
+    const auto find_set = nodeset_ids.find(_mesh->getBoundaryID(nodeset_id));
+    if (find_set == nodeset_ids.end())
+      paramError("nodeset_list",
+                 "The nodeset ",
+                 nodeset_id,
+                 " provided for the contact action is not in the mesh. "
+                 "Please revise your input file.");
+
+    // Accumulate components for center of gravity computations
+    Point cog(0.0, 0.0, 0.0);
+    std::vector<dof_id_type> nodelist = _mesh->getNodeList(_mesh->getBoundaryID(nodeset_id));
+
+    for (const auto & it : nodelist)
+    {
+      const Node * const node = _mesh->queryNodePtr(it);
+      cog += *node;
+    }
+
+    cog(0) /= nodelist.size();
+    cog(1) /= nodelist.size();
+    cog(2) /= nodelist.size();
+
+    nodeset_list_cog.emplace_back(nodeset_id, cog);
+  }
+
+  // Vectors of distances for each pair
+  std::vector<std::pair<std::pair<BoundaryName, BoundaryName>, Real>> pairs_distances;
+
+  // Assign distances to identify nearby pairs.
+  for (std::size_t i = 0; i < nodeset_list_cog.size() - 1; i++)
+    for (std::size_t j = i + 1; j < nodeset_list_cog.size(); j++)
+    {
+      const Point & distance_vector = nodeset_list_cog[i].second - nodeset_list_cog[j].second;
+
+      if (nodeset_list_cog[i].first != nodeset_list_cog[j].first)
+      {
+        const Real distance = distance_vector.norm();
+        const std::pair pair = std::make_pair(nodeset_list_cog[i].first, nodeset_list_cog[j].first);
+        pairs_distances.emplace_back(std::make_pair(pair, distance));
+      }
+    }
+
+  Real minimum_representative_distance(std::numeric_limits<Real>::max());
+
+  // Iterate over pairs-distance and find a represenh
+  for (const auto & pair_distance : pairs_distances)
+    if (pair_distance.second > TOLERANCE && pair_distance.second < minimum_representative_distance)
+      minimum_representative_distance = pair_distance.second;
+
+  // Apply a tolerance to the minimum distance to avoid missing pairs
+  minimum_representative_distance *= 1.2;
```

Comment:
@bwspenc: Should we require user input here to gauge whether a reference distance would grant the generation of a new contact pair? I believe that for the case we have in mind is not strictly necessary, but may be good for other applications?

Reply:
Per our discussion, we will ask the user to provide a reference distance that will include some tolerance to capture computation error.